x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/CrossClusterSecurity.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.transport;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.settings.ClusterSettings;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.transport.TcpTransport;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeSet;
+
+import static org.elasticsearch.transport.RemoteClusterService.REMOTE_CLUSTER_AUTHORIZATION;
+
+/**
+ * Load/reload API Keys from cluster settings.
+ * Injected into SecurityServerTransportInterceptor.
+ * Latest API Key value needs to be used in header when sending TransportRequest to remote cluster nodes.
+ */
+public class CrossClusterSecurity {
+
+    private static final Logger LOGGER = LogManager.getLogger(CrossClusterSecurity.class);
+
+    private final Map<String, String> apiKeys = ConcurrentCollections.newConcurrentMap();
+
+    public CrossClusterSecurity(final Settings settings, final ClusterSettings clusterSettings) {
+        if (TcpTransport.isUntrustedRemoteClusterEnabled()) {
+            this.setApiKeys(REMOTE_CLUSTER_AUTHORIZATION.getAsMap(settings));
+            clusterSettings.addAffixMapUpdateConsumer(REMOTE_CLUSTER_AUTHORIZATION, this::setApiKeys, (k, v) -> {});
+        }
+    }
+
+    public String getApiKey(final String clusterAlias) {
+        if (TcpTransport.isUntrustedRemoteClusterEnabled()) {
+            return this.apiKeys.get(clusterAlias);
+        } else {
+            return null;
+        }
+    }
+
+    private void setApiKeys(final Map<String, String> newClusterAliasApiKeyMap) {
+        if (TcpTransport.isUntrustedRemoteClusterEnabled()) {
```

Comment:
This map being passed in here is an instance of IdentityHashMap. Why? This breaks Map.equals and Map.get below. I traced it to Settings.java. Why are new cluster settings passed in as IdentityHashMap, instead of a regular HashMap?

https://github.com/elastic/elasticsearch/blob/db7b0cbbbdbebd1b40eddb5045c490d8999febfe/server/src/main/java/org/elasticsearch/common/settings/Setting.java#L928